### DISCUSSION TITLE
v3 - is migration to reactivesearch API mandatory? (and architecture questions)

### DISCUSSION THREAD

**Discussion by alancwoo** [appbaseio/reactivesearch#2197] [Q&A] [answered]
@siddharthlatest I was curious about the 3.x/Vue3 version - does the [removal of FE query generation](https://docs.reactivesearch.io/docs/reactivesearch/vue/advanced/migration/#removal-of-fe-query-generation-1) mean this library is only compatible with Reactivesearch API and can no longer communicate directly to an elasticsearch node? _(my understanding is the answer is yes the library can no longer communicate with a vanilla elasticsearch node)_

If this is the case, I was wondering how exactly everything fits together - does this mean that an independent reactivesearch cluster must be spun up and all the data imported from the original elasticsearch node, or does the reactivesearch-api act as a middle layer between the frontend app and the original elasticsearch node? _(my understanding is also yes that a reactivesearch-api server has to be spun up and the data imported)_

Also, how does one go about keeping the reactivesearch-api database synchronized/up to date?

**Comment by siddharthlatest** [comment, marked as answer]
@alancwoo Starting Vue 3.0 and React 4.0 releases (they're still in RC), as you shared, the library is only compatible with ReactiveSearch API and can no longer communicate directly to an Elasticsearch cluster.

**I will share more on the why:**
- The ReactiveSearch UI kits will support more search engines than just Elasticsearch going forward, and the use of ReactiveSearch API (a declarative wrapper of passing the search intent) is what makes this possible
- Exposing Elasticsearch DSL to the frontend (unless it's trusted users, on the intranet) creates a DoS vector
- We are aiming to build a sustainable and fair business model based on users that use ReactiveSearch UI kits, where we can continue to invest in the open-source development.

I will be writing a doc on how everything fits together fairly soon as we're approaching a stable release for both major versions, and more importantly streamline the deployment. There will be two modes supported:
- Host your Elasticsearch or OpenSearch instance on ReactiveSearch cloud: This comes provisioned with ReactiveSearch API as the gateway and is the recommended choice for someone starting out
- Connect your Elasticsearch or OpenSearch instance to ReactiveSearch cloud: When you already have an existing Elasticsearch cluster, you can run ReactiveSearch cloud as a reverse proxy to this instance

along with self-host (with Docker) and AWS marketplace deployment option for some rare use-cases.

> Also, how does one go about keeping the reactivesearch-api database synchronized/up to date?

For the cloud and AWS deployments, this is straight forward. If self-hosting, you will need to udpate the Docker image versions. We are maintaining release artifacts on this repository: https://github.com/appbaseio/reactivesearch-api/releases

Thank you for being a long-term user, and I hope you and others continue to use ReactiveSearch as we transition to the upcoming major releases. 🙂

A streamlined process and a doc on this is coming out very soon!

**Reply by alancwoo** [reply]
Thanks for the clear response, this is very helpful to know. Look forward to the updates as things move ahead!
